### PR TITLE
Inline asm types

### PR DESCRIPTION
**Pull request by Jasper-Bekkers**
This most likely supersedes #372.

Various SPIRV instructions have operands that can require more operands when they're specified. Image instructions are one example, `OpDecorate` is another.

Todo:
 - [x] Add tests

### CODE REVIEW

**Review by eddyb — APPROVED** [review]

File: crates/rustc_codegen_spirv/src/linker/capability_computation.rs [hunk 1]
```diff
@@ -104,28 +72,36 @@ fn operand_required_extensions(op: &Operand) -> &'static [&'static str] {
...
                 .chain(additional_extensions(module, inst))
         })
         .copied()
         .collect();
+
+    set.extend(
+        module
+            .all_inst_iter()
+            .flat_map(|inst| inst.operands.iter().flat_map(|op| op.required_extensions())),
+    );
```
Comment (khyperia) [inline crates/rustc_codegen_spirv/src/linker/capability_computation.rs:103]: Why did you split this out to its own thing? It's less efficient this way

File: crates/rustc_codegen_spirv/src/linker/capability_computation.rs [hunk 2]
```diff
@@ -104,28 +72,36 @@ fn operand_required_extensions(op: &Operand) -> &'static [&'static str] {
 // It says no. We strip it. Things explode.
 // So, this function is to encode any special version-specific rules that aren't in rspirv.
 fn additional_extensions(module: &Module, inst: &Instruction) -> &'static [&'static str] {
-    if module.header.as_ref().unwrap().version() < (1, 5)
-        && inst.class.opcode == Op::Capability
-        && inst.operands[0].unwrap_capability() == Capability::VulkanMemoryModel
-    {
-        &["SPV_KHR_vulkan_memory_model"]
+    let version = module.header.as_ref().unwrap().version();
```
Comment (khyperia) [inline crates/rustc_codegen_spirv/src/linker/capability_computation.rs:75]: Probably good to move this let inside the if statement, as this is really hot code

File: crates/rustc_codegen_spirv/src/linker/capability_computation.rs [hunk 3]
```diff
@@ -104,28 +72,36 @@ fn operand_required_extensions(op: &Operand) -> &'static [&'static str] {
...
+    if inst.class.opcode == Op::Capability {
+        match inst.operands[0].unwrap_capability() {
+            Capability::VulkanMemoryModel if version < (1, 5) => &["SPV_KHR_vulkan_memory_model"],
+            Capability::RuntimeDescriptorArray => &["SPV_EXT_descriptor_indexing"],
```
Comment (khyperia) [inline crates/rustc_codegen_spirv/src/linker/capability_computation.rs:79]: I believe this is incorrect, it should be `if version < (1, 5)`

File: crates/rustc_codegen_spirv/src/builder/spirv_asm.rs [hunk 4]
```diff
@@ -436,8 +492,8 @@ impl<'cx, 'tcx> Builder<'cx, 'tcx> {
         }
         if tokens.next().is_some() {
             self.tcx.sess.err(&format!(
-                "too many operands to instruction: {}",
-                instruction.class.opname
+                "too many operands to instruction: {} {:?}",
+                instruction.class.opname, instruction
```
Comment (khyperia) [inline crates/rustc_codegen_spirv/src/builder/spirv_asm.rs:496]: The debug impl includes the opname, probably best to print it only once (but also the debug impl is pretty gross, do we really want to include it in an error message?)

File: crates/rustc_codegen_spirv/src/builder/spirv_asm.rs [hunk 5]
```diff
@@ -256,10 +256,47 @@ impl<'cx, 'tcx> Builder<'cx, 'tcx> {
...
+            Op::Extension => {
+                self.emit_global()
+                    .extension(inst.operands[0].unwrap_literal_string());
+                return;
+            }
+            Op::Capability => {
+                self.emit_global()
+                    .capability(inst.operands[0].unwrap_capability());
+                return;
+            }
```
Comment (khyperia) [inline crates/rustc_codegen_spirv/src/builder/spirv_asm.rs:292]: Why are these needed?

**Review by khyperia — APPROVED** [review]

**Review by khyperia — CHANGES_REQUESTED** [review]
hit the wrong button, sorry, need some morning tea :(

File: crates/rustc_codegen_spirv/src/builder/spirv_asm.rs (hunk 5 above)
Comment (eddyb) [reply crates/rustc_codegen_spirv/src/builder/spirv_asm.rs:292]: My guess was that `emit_global` helped with avoiding inserting these into the current block in a function, but maybe `rspirv` handles that already?

Or, wait, should `OpExtension`/`OpCapability` never be manually specified in `asm!`, and always deduced from other instructions?

File: crates/rustc_codegen_spirv/src/builder/spirv_asm.rs (hunk 4 above)
Comment (eddyb) [reply crates/rustc_codegen_spirv/src/builder/spirv_asm.rs:496]: TIL `opname`, is that like `format!("{:?}", instruction.class.opcode)`?

File: crates/rustc_codegen_spirv/src/linker/capability_computation.rs (hunk 1 above)
Comment (Jasper-Bekkers) [reply crates/rustc_codegen_spirv/src/linker/capability_computation.rs:103]: Iterator sadness - `required_extensions` became something that returns a Vec<> instead of a &'static slice now and I could get the chain calls to match up anymore.

File: crates/rustc_codegen_spirv/src/builder/spirv_asm.rs (hunk 5 above)
Comment (Jasper-Bekkers) [reply crates/rustc_codegen_spirv/src/builder/spirv_asm.rs:292]: They weren't supported before and were needed in my test sample, I can move them to a different PR if you feel the need.

@eddyb `OpCapability` and `OpExtension` cant always be deduced from the SPIR-V opcode unfortunately. In my testing I needed `SPV_EXT_descriptor_indexing` and `RuntimeDescriptorArray` to be able to index into an `OpTypeRuntimeArray` of images, which otherwise just isn't allowed.

File: crates/rustc_codegen_spirv/src/builder/spirv_asm.rs (hunk 4 above)
Comment (Jasper-Bekkers) [reply crates/rustc_codegen_spirv/src/builder/spirv_asm.rs:496]: This is debug left-over that I can remove.

**Review by khyperia — APPROVED** [review]
LGTM pending tests if you'd like to write those! :tada:

**Comment by Jasper-Bekkers** [comment]
Adding a test for `OpImageSampleProjExplicitLod` also to show how to use BitEnums.
